libs/src/sdk/api/tracks/TracksApi.ts

```diff
@@ -75,44 +81,102 @@ export class TracksApi extends TracksApiWithoutStream {
       throw new Error(
         `Metadata object is missing values: ${metadataMissingValues}`
       )
     }
 
     // Upload track audio and cover art to storage node
-    const updatedMetadata = await this.storage.uploadTrackAudioAndCoverArt(
-      trackFile,
-      coverArtFile,
-      metadata,
-      onProgress
-    )
+    const [audioResp, coverArtResp] = await Promise.all([
+      retry3(
+        async () =>
+          await this.storage.uploadFile({
+            file: trackFile,
+            onProgress,
+            template: 'audio'
+          }),
+        (e) => {
+          console.log('Retrying uploadTrackAudio', e)
+        }
+      ),
+      retry3(
+        async () =>
+          await this.storage.uploadFile({
+            file: coverArtFile,
+            onProgress,
+            template: 'img_square'
+          }),
+        (e) => {
+          console.log('Retrying uploadTrackCoverArt', e)
+        }
+      )
+    ])
+
+    // Update metadata to include uploaded CIDs
+    const updatedMetadata = {
+      ...metadata,
+      track_segments: [],
+      track_cid: audioResp.results['320'],
+      download: metadata.download?.is_downloadable
+        ? {
+            ...metadata.download,
+            cid: metadata.track_cid
+          }
+        : metadata.download,
+      cover_art_sizes: coverArtResp.id,
+      duration: parseInt(audioResp.probe.format.duration, 10)
+    }
 
-    // TODO: Integrate with wallet api
-    // TODO: Write metadata to chain
-    // const trackId = await this._generateTrackId()
-    // const response = await this.contracts.EntityManagerClient!.manageEntity(
-    //   ownerId,
-    //   EntityManagerClient.EntityType.TRACK,
-    //   trackId,
-    //   EntityManagerClient.Action.CREATE,
-    //   JSON.stringify({ cid: updatedMetadata.track_cid, data: updatedMetadata })
-    // )
-    // const txReceipt = response.txReceipt
+    // Write metadata to chain
+    const metadataCid = await generateMetadataCidV1(updatedMetadata)
+    const trackId = await this.generateTrackId()
+    const response = await this.entityManager.manageEntity({
+      userId: artistId,
+      entityType: EntityType.TRACK,
+      entityId: trackId,
+      action: Action.CREATE,
+      metadata: JSON.stringify({
+        cid: metadataCid.toString(),
+        data: updatedMetadata
+      }),
+      walletApi: this.walletApi
+    })
+    const txReceipt = response.txReceipt
 
     return {
-      // blockHash: txReceipt.blockHash,
-      // blockNumber: txReceipt.blockNumber,
-      // trackId,
+      blockHash: txReceipt.blockHash,
+      blockNumber: txReceipt.blockNumber,
+      trackId,
       transcodedTrackCID: updatedMetadata.track_cid,
       error: false
     }
   }
+
+  private async generateTrackId() {
+    const response = await this.request({
+      path: `/tracks/unclaimed_id`,
+      method: 'GET',
+      headers: {},
+      query: { noCache: Math.floor(Math.random() * 1000).toString() }
```

Comment:
I guess just to prevent the response from being cached, copied from here https://github.com/AudiusProject/audius-protocol/blob/main/libs/src/services/discoveryProvider/requests.ts#L807